File: lesson4/lesson4/ViewController.m
//
//  ViewController.m
//  lesson4
//
//  Created by ityike on 2016/12/5.
//  Copyright © 2016年 袁 峰. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()


// @property (weak, nonatomic) IBOutlet UIView *redView;

@property (weak, nonatomic) UIView *redView;

@property (weak, nonatomic) UILabel *label;

@end

/**
 尽量使用tag
 tag的性能差
 用多了容易乱
 */

@implementation ViewController

- (void)viewDidLoad {

    [super viewDidLoad];
    
    // 根据tag拿到对应的view
    UIView *redView = [self.view viewWithTag:1];
    self.redView = redView;
    
    // 1 创建UISwitch对象
    UISwitch *sw = [[UISwitch alloc] init];
    
    // 2 加载控制器的view
    [self.view addSubview:sw];
    
    // 3 创建UISWitch对象
    UISwitch *sw1 = [[UISwitch alloc] init];
    
    // 4 加到红色的view
    [redView addSubview:sw1];
    
    // 5 创建一个选项卡对象
    UISegmentedControl *sg = [[UISegmentedControl alloc] initWithItems:@[@"哈哈哈", @"嘻嘻"]];
    // 6 加到红色的view
    [redView addSubview:sg];
    
    // 7 移除
    [sg removeFromSuperview];
    [sw1 removeFromSuperview];
    
    
    // 创建UILabel对象
    UILabel *label = [[UILabel alloc] init];
    // 设置frame
    label.frame = CGRectMake(100, 100, 100, 60);
    // 设置背景颜色
    label.backgroundColor = [UIColor yellowColor];
    // 添加到view中
    [self.view addSubview:label];
    
    self.label = label;
}

- (IBAction)remove {
    [self.redView removeFromSuperview];
}

- (IBAction)bounds {
    // 改变尺寸
    self.label.bounds = CGRectMake(0, 0, 200, 10);
}

- (IBAction)center {
    // 改变位置
    //self.label.center = CGPointMake(100, 100);
    // 显示在view的最中间
    self.label.center = CGPointMake(self.view.frame.size.width / 2, self.view.frame.size.height / 2);
}

- (IBAction)changeframe {
   // 方式1
//    self.label.frame = CGRectMake(200, 100, 200, 160);
    
    // 方式2
    CGRect frame = self.label.frame;
    frame.origin.x += 100;
    frame.origin.y += 100;
    frame.size.width += 100;
    frame.size.height -= 30;
    self.label.frame = frame;
    
}

// 只要控件有父控件，就一定能移除。
- (void)viewDidAppear:(BOOL)animated{
    [super viewDidAppear:animated];
    //[self.view removeFromSuperview];
}


@end

File: 第六天/懒加载/懒加载/ViewController.swift
//
//  ViewController.swift
//  懒加载
//
//  Created by ityike on 2016/12/19.
//  Copyright © 2016年 袁 峰. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
    
//    var label: DemoLabel?
    
    // 初始化label 并且分配了空间, 会提前创建
    // 在移动开发中，延迟加载减少内存消耗
    // 懒加载 - lazy
    
    // 延迟创建
    // 最大的好处 -> 解除解包的烦恼
    //lazy var label: DemoLabel = DemoLabel()
    
    // 懒加载本质上是一个闭包
    // 完整写法：
    // {} 包装代码 () 执行代码
    // 日常开发不这样写！
    // 闭包中如果出现self. 还需要注意循环引用。
    lazy var label = { () -> DemoLabel in
        let label1 = DemoLabel()
        // 设置 label 的属性
        return label1
    }()

    override func viewDidLoad() {
        super.viewDidLoad()
        setUI()
    }
    
    
    private func setUI() {
//        // 创建控件
//        label = DemoLabel()
        
        // 添加到视图
        // ! - 解包，为了参与计算，addsubview 用subview 数组记录控件，数组中不允许插入nil
        // ? - 可选解包  调用方法，如果为nil，不调用方法，但是不能参与计算
        view.addSubview(label)
        
        label.text = "hello"
        label.sizeToFit()
        label.center = view.center
    }


}

